File: python/reverse_vowels_of_a_string.py
"""
#slower solution: 
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = { "a", "e", "i", "o", "u" }
        #keep track of vowels we find, in order
        vowel_stack = []
        #keep track of string indices that have vowels, in order
        vowel_indices = []
        #store the result as a list since strings are not mutable
        result = [None for i in range(len(s))]
        for i, ch in enumerate(s):
            if ch.lower() in vowels:
                vowel_stack.append(ch)
                vowel_indices.append(i)
            else:
                #if the character isn't a vowel, we can add it in its current index to the result
                result[i] = ch
        for i in vowel_indices:
            result[i] = vowel_stack.pop()
        return "".join(result)
"""

class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = { "a", "e", "i", "o", "u" }
        # use a list since its mutable
        s_as_list = list(s)
        left_pointer, right_pointer = -1, len(s)
        # the left and right pointers search for vowels one at a time and switch them
        while right_pointer > left_pointer:
            left_pointer += 1
            while left_pointer < len(s) and s_as_list[left_pointer].lower() not in vowels:
                left_pointer += 1
            right_pointer -= 1
            while right_pointer >= 0 and s_as_list[right_pointer].lower() not in vowels:
                right_pointer -= 1
            if right_pointer > left_pointer:
                s_as_list[left_pointer], s_as_list[right_pointer] = s_as_list[right_pointer], s_as_list[left_pointer]
        return "".join(s_as_list)

File: java/reverseVowelsOfAString.java

//slow solution:

// class Solution {
//     public String reverseVowels(String s) {
//         Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u'));
//         List<Character> vowelStack = new ArrayList<>();
//         List<Integer> vowelIndices = new ArrayList<>();
//         char[] result = new char[s.length()];
//         for (int i = 0; i < s.length(); i++) {
//             char c = s.charAt(i);
//             if (vowels.contains(Character.toLowerCase(c))) {
//                 vowelStack.add(c);
//                 vowelIndices.add(i);
//             } else {
//                 result[i] = c;
//             }
//         }
//         for (int i = 0; i < vowelIndices.size(); i++) {
//             int resultIndex = vowelIndices.get(i);
//             char nextChar = vowelStack.get(vowelStack.size() - 1 - i);
//             result[resultIndex] = nextChar;
//         }
//         return new String(result);
//     }
// }


class Solution {
    public String reverseVowels(String s) {
        Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u'));
        //string builder will allow us to change characters of the string in place
        StringBuilder sb = new StringBuilder(s);
        int leftPointer = -1;
        int rightPointer = s.length();
        //the left and right pointers will traverse the string and fine one vowel at a time and switch them
        while (rightPointer > leftPointer) {
            do {
                leftPointer++;
            }
            while (leftPointer < s.length() && !vowels.contains(Character.toLowerCase(sb.charAt(leftPointer))));
            do {
                rightPointer--;
            }
            while (rightPointer >= 0 && !vowels.contains(Character.toLowerCase(sb.charAt(rightPointer))));
            if (rightPointer > leftPointer) {
                char c = sb.charAt(leftPointer);
                sb.setCharAt(leftPointer, sb.charAt(rightPointer));
                sb.setCharAt(rightPointer, c);
            }  
        }
        return sb.toString();
    }
}